File: AccuPay/Domain/TimeEntry/Overtime.vb
Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema
Imports AccuPay.Utilities.Extensions

<Table("employeeovertime")>
Public Class Overtime

    Public Const StatusApproved = "Approved"

    Public Const StatusPending = "Pending"

    Public Const DefaultType = "Overtime"

    <Key>
    <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
    Public Overridable Property RowID As Integer?

    Public Overridable Property OrganizationID As Integer?

    <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
    Public Overridable Property Created As DateTime

    Public Overridable Property CreatedBy As Integer?

    <DatabaseGenerated(DatabaseGeneratedOption.Computed)>
    Public Overridable Property LastUpd As DateTime?

    Public Overridable Property LastUpdBy As Integer?

    Public Overridable Property EmployeeID As Integer?

    <Column("OTType")>
    Public Overridable Property Type As String

    Public Overridable Property OTStartTime As TimeSpan?

    Public Overridable Property OTEndTime As TimeSpan?

    Public Overridable Property OTStartDate As Date

    Public Overridable Property OTEndDate As Date

    <Column("OTStatus")>
    Public Overridable Property Status As String

    <NotMapped>
    Public Overridable Property Start As Date?

    <NotMapped>
    Public Overridable Property [End] As Date?

    Public Overridable Property Reason As String

    Public Overridable Property Comments As String

    Public Sub New()
        Status = StatusPending
    End Sub

    <NotMapped>
    Public Property OTStartTimeFull As Date?
        Get
            'Using Nothing as output on ternary operator does not work
            If OTStartTime Is Nothing Then
                Return Nothing
            Else
                Return OTStartDate.Date.ToMinimumHourValue.Add(OTStartTime.Value)
            End If
        End Get
        Set(value As Date?)
            'Using Nothing as output on ternary operator does not work
            If value Is Nothing Then
                OTStartTime = Nothing
            Else
                OTStartTime = value?.TimeOfDay
            End If
        End Set
    End Property

    <NotMapped>
    Public Property OTEndTimeFull As Date?
        Get
            'Using Nothing as output on ternary operator does not work
            If OTEndTime Is Nothing Then
                Return Nothing
            Else
                Return OTEndDate.Date.ToMinimumHourValue.Add(OTEndTime.Value)
            End If
        End Get
        Set(value As Date?)
            'Using Nothing as output on ternary operator does not work
            If value Is Nothing Then
                OTEndTime = Nothing
            Else
                OTEndTime = value?.TimeOfDay
            End If
        End Set
    End Property

    Public Function Validate() As String

        If OTStartTime Is Nothing Then

            Return "Start Time cannot be empty."
        End If

        If OTEndTime Is Nothing Then

            Return "End Time cannot be empty."
        End If

        If OTStartDate.Date.Add(OTStartTime.Value.StripSeconds) >= OTEndDate.Date.Add(OTEndTime.Value.StripSeconds) Then

            Return "Start date and time cannot be greater than or equal to End date and time."

        End If

        If Not {StatusPending, StatusApproved}.Contains(Status) Then

            Return "Status is not valid."

        End If

        'Means no error
        Return Nothing

    End Function

End Class

File: AccuPay.Core/Entities/Overtime.cs
using AccuPay.Core.Interfaces;
using AccuPay.Utilities.Extensions;
using System;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Core.Entities
{
    [Table("employeeovertime")]
    public class Overtime : EmployeeDataEntity, IPayrollEntity
    {
        public const string StatusApproved = "Approved";

        public const string StatusPending = "Pending";

        [Column("OTType")]
        public string Type { get; set; }

        public TimeSpan? OTStartTime { get; set; }

        public TimeSpan? OTEndTime { get; set; }

        public DateTime OTStartDate { get; set; }

        // TODO: make this readonly. Domain methods should only be the one to set this.
        public DateTime OTEndDate { get; set; }

        [Column("OTStatus")]
        public string Status { get; set; }

        public string Reason { get; set; }

        public string Comments { get; set; }

        [ForeignKey("EmployeeID")]
        public virtual Employee Employee { get; set; }

        public DateTime? PayrollDate => OTStartDate;

        private Overtime()
        {
            Status = StatusPending;
        }

        [NotMapped]
        public DateTime? OTStartTimeFull
        {
            get => OTStartTime == null ?
                (DateTime?)null :
                OTStartDate.Date.ToMinimumHourValue().Add(OTStartTime.Value);

            set => OTStartTime = value == null ? null : value?.TimeOfDay;
        }

        [NotMapped]
        public DateTime? OTEndTimeFull
        {
            get => OTEndTime == null ?
                (DateTime?)null :
                OTEndDate.Date.ToMinimumHourValue().Add(OTEndTime.Value);

            set => OTEndTime = value == null ? null : value?.TimeOfDay;
        }

        public void UpdateEndDate()
        {
            if (OTStartTime == null || OTEndTime == null)
            {
                OTEndDate = OTStartDate;
            }
            else
            {
                OTEndDate = OTEndTime < OTStartTime ? OTStartDate.AddDays(1) : OTStartDate;
            }
        }

        public static Overtime NewOvertime(IShift shift, int organizationId)
        {
            return NewOvertime(
                organizationId: organizationId,
                employeeId: shift.EmployeeId.Value,
                startDate: shift.Date,
                startTime: null,
                endTime: null,
                status: StatusApproved);
        }

        public static Overtime NewOvertime(
            int organizationId,
            int employeeId,
            DateTime startDate,
            TimeSpan? startTime,
            TimeSpan? endTime,
            string status = StatusApproved,
            string reason = null,
            string comments = null)
        {
            var overtime = new Overtime()
            {
                OrganizationID = organizationId,
                EmployeeID = employeeId,
                OTStartDate = startDate,
                OTStartTime = startTime,
                OTEndTime = endTime,
                Status = status,
                Reason = reason,
                Comments = comments
            };

            overtime.UpdateEndDate();

            return overtime;
        }
    }
}
